forwardproxy_udp.go

```diff
@@ -0,0 +1,593 @@
+		}
+		if r.Header.Get("Upgrade") != requestProtocol {
+			return "", fmt.Errorf("unexpected Upgrade: %s", r.Proto)
+		}
+	case 2:
+		// copy from https://github.com/caddyserver/caddy/blob/master/modules/caddyhttp/reverseproxy/reverseproxy.go#L414-L428
+		// check HTTP2 request
+		if r.Method != http.MethodConnect {
+			return "", fmt.Errorf("expected CONNECT request, got %s", r.Method)
+		}
+		if r.Header.Get(":protocol") != requestProtocol {
+			return "", fmt.Errorf("unexpected protocol: %s", r.Header.Get(":protocol"))
+		}
+	case 3:
+		// copy from https://github.com/quic-go/masque-go/blob/master/request.go#L47-L119
+		// and remove ckecking host
+		if r.Method != http.MethodConnect {
+			return "", fmt.Errorf("expected CONNECT request, got %s", r.Method)
+		}
+		if r.Proto != requestProtocol {
+			return "", fmt.Errorf("unexpected protocol: %s", r.Proto)
+		}
+	default:
+		return "", fmt.Errorf("unexpected HTTP version: %v", r.ProtoMajor)
+	}
+
+	// The capsule protocol header is optional, but if it's present,
+	// we need to validate its value.
+	capsuleHeaderValues, ok := r.Header[capsuleProtocolHeader]
+	if ok {
+		item, err := httpsfv.UnmarshalItem(capsuleHeaderValues)
+		if err != nil {
+			return "", fmt.Errorf("invalid capsule header value: %s", capsuleHeaderValues)
+		}
+		if v, ok := item.Value.(bool); !ok {
+			return "", fmt.Errorf("incorrect capsule header value type: %s", reflect.TypeOf(item.Value))
+		} else if !v {
+			return "", fmt.Errorf("incorrect capsule header value: %t", item.Value)
+		}
+	}
+
+	// check request as UDP bind
+	var isUDPBind bool
+	connectUDPBindHeaderValues, ok := r.Header[connectUDPBindHeader]
+	if ok {
+		item, err := httpsfv.UnmarshalItem(connectUDPBindHeaderValues)
+		if err != nil {
+			return "", fmt.Errorf("invalid bind header value: %s", connectUDPBindHeaderValues)
+		}
+		if v, ok := item.Value.(bool); !ok {
+			return "", fmt.Errorf("incorrect bind header value type: %s", reflect.TypeOf(item.Value))
+		} else if !v {
+			return "", fmt.Errorf("incorrect bind header value: %t", item.Value)
+		}
+		isUDPBind = true
+	}
+
+	match := template.Match(r.URL.String())
+	targetHost := func(s string) string { return strings.ReplaceAll(s, "%3A", ":") }(match.Get("target_host").String())
+	targetPortStr := match.Get("target_port").String()
+	if targetHost == "" || targetPortStr == "" {
+		return "", fmt.Errorf("expected target_host and target_port")
+	}
+	if isUDPBind && targetHost == "*" && targetPortStr == "*" {
+		return "*", nil
+	}
+	// IPv6 addresses need to be enclosed in [], otherwise resolving the address will fail.
+	if strings.Contains(targetHost, ":") {
+		targetHost = "[" + targetHost + "]"
+	}
+	targetPort, err := strconv.Atoi(targetPortStr)
+	if err != nil {
+		return "", fmt.Errorf("failed to decode target_port: %w", err)
+	}
+	return Request(net.JoinHostPort(targetHost, strconv.Itoa(targetPort))), nil
+}
+
+func (h *Handler) checkACL(hostPort string) (bool, error) {
```

Comment:
Seems to have duplicated forwardproxy.go.

I think must of this should be embedded along side the logic of tcp/http.

Reply:
Done